### DISCUSSION TITLE
OpenTelemetry-dotnet: Facing Duplicate Data Issues with UpDownCounters and ObservableGauges

### DISCUSSION THREAD

**Discussion by abhishek8** [open-telemetry/opentelemetry-dotnet#5753] [Q&A] [answered]
Hi, I have just started using OpenTelemetry with .NET. I have a use case where I need to log specific data points, such as success and failure counts in the application. To avoid data duplication, I am using Delta temporality preference. While I’ve been successful with regular counters, I’m still encountering duplicate data issues with UpDownCounters and ObservableGauges. Can anyone explain how to resolve this?

Below sample code for the same,
```
using var meterProvider = Sdk.CreateMeterProviderBuilder()
    .AddMeter("MyCompany.MyProduct.MyLibrary")
    .AddConsoleExporter((options, readerOptions) =>
    {
        readerOptions.TemporalityPreference = MetricReaderTemporalityPreference.Delta;
    })
    .Build();

var abc = new Main();
abc.Execute();

internal class Main
{
    private readonly Meter _meter = new("MyCompany.MyProduct.MyLibrary");

    public void Execute()
    {
        Console.WriteLine("Started...");

        _meter.CreateCounter<long>("MyFruitCounter_v1")
            .Add(3, new("name", "apple"), new("color", "red"), new("uid", "9161"));

        _meter.CreateUpDownCounter<long>("MyFruitGauge_v1")
            .Add(3, new("name", "apple"), new("color", "red"), new("uid", "9161"));

        Console.WriteLine("Logged values!");

        while (!Console.KeyAvailable)
        {
            Thread.Sleep(1000);
        }
    }
}
```
<img width="658" alt="OpenTelemetry_Duplicate" src="https://github.com/user-attachments/assets/e0300817-dbd1-4a35-91d9-ff72294b1d9d">

**Comment by cijothomas** [comment]
> To avoid data duplication, I am using Delta temporality preference. 

I am not sure what does this mean. What is the connection between Temporality and data duplication?

**Comment by cijothomas** [comment]
>  I’m still encountering duplicate data issues with UpDownCounters and ObservableGauges.

Can you elaborate on what is "duplicate" data here? 
(Also, the shared screenshot only has Counters and UpDownCounters, and no ObservableGauges!)

**Reply by abhishek8** [reply]
By "duplicate" I mean same metrices is getting logged multiple times. I have logged them only once so they should appear in console only once. At above example "MyFruitCounter_v1" appears only time but "MyFruitGauge_v1" keeps getting logged again and again.

**Reply by cijothomas** [reply, marked as answer]
Thanks for clarifying what you meant by "duplicate". What you see is expected behavior. For UpDownCounters, there is no true delta support today. Even when you specify delta, UpDownCounters still take Cumulative only, and hence they are reported with each export!

(it is possible to new provide a feature to control temporality more precisely. please open a feature request as a Github issue, stating you need support for true Delta UpDownCounters.)

**Reply by abhishek8** [reply]
Thanks for answering. One more thing if I may ask, is there any workaround at present in OpenTelemetry to log specific events just as we do with Influx. My scenario is to monitor success and failure counts in batch processing.

**Reply by cijothomas** [reply]
> log specific events just as we do with Influx

I am not familiar with that. If you can describe details, I can help.

**Reply by abhishek8** [reply]
I simply need the right set of values coming in for "MyFruitGauge_v1" each time so that I can plot them as below,

<img width="446" alt="image" src="https://github.com/user-attachments/assets/e75dd445-8097-47e4-aa9b-5c0d3df7eb3b">
